stm32: pwm driver enables complementary output for timer channel

FRASTM opened the pull request:
with this PR, the complementary PWM output is enabled when the pin is choosen in the DTS of the target

Using a newly defined PWM_COMPLEMENTARY flag in the target board DTS as follows:
```
		red_pwm_led: red_pwm_led {
			pwms = <&pwm1 2 4 (PWM_POLARITY_NORMAL | PWM_COMPLEMENTARY)>;
		};
```

- At this time, only used by the stm32 mcus
- Does not impact the PWM input capture.


Fixes https://github.com/zephyrproject-rtos/zephyr/issues/40657

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: pwm: stm32 can initialize the PWM complementary output

If the dts defines the PWM complementary output, then the OCN
must be init in place of the OC state and polarity.
This is an exclusive setting for this pin.
The channel in LL_TIM_OC_SetPolarity can be the complementary one.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Tested on the nucleo_l4r5zi target where the PB 14 pin which is for the red Led is a output of a PWM signal from
```
-  Timer15 channel 1, when
-  		red_pwm_led: red_pwm_led {
			pwms = <&pwm15 1 4 PWM_POLARITY_NORMAL>;
		};
```

- or Timer 1 channel 2 N (complementary output), when 
```
- 		red_pwm_led: red_pwm_led {
			pwms = <&pwm1 2 4 PWM_POLARITY_NORMAL>;
		};
```

gmarull reviewed (changes requested):
pinctrl information is not meant to be used by drivers directly. If you want PWM to enable complementary output, you should add an additional flag at the pwm node or, propose a new PWM flag. Then user is responsible to set pinctrl appropriately.

erwango commented:
> pinctrl information is not meant to be used by drivers directly. If you want PWM to enable complementary output, you should add an additional flag at the pwm node or, propose a new PWM flag. Then user is responsible to set pinctrl appropriately.

IIUC this means that we should be able to identify a PWM channel to a pin.
For single channel PWM configuration this is fine, but what about multiple channel configuration, as for instance:
```
	pwm4: pwm {
		status = "okay";
		pinctrl-0 = <&tim4_ch1_pd12
			     &tim4_ch2n_pd13
			     &tim4_ch3n_pd14
			     &tim4_ch4_pd15>;
		pinctrl-names = "default";
	};
```
Does it mean that we can rely on the order of the pins established by the user in pins array ?
(Current work is anyway not sufficient to achieve this anyway as we miss channel to pin matching, but I'd like to understand what are the possibilities)

gmarull commented:
One option would be to do something like this:

```
red_pwm_led: red_pwm_led {
    pwms = <&pwm12 1 4 (PWM_POLARITY_NORMAL | PWM_COMPLEMENTARY)>;
}; 
```

gmarull commented:
Or, another option (not tested):

```
#define STM32_PWM_CH(ch) (ch)
#define STM32_PWM_CHN(ch) ((ch) << 3U)

red_pwm_led: red_pwm_led {
    /* use regular output */
    pwms = <&pwm12 STM32_PWM_CH(1) 4 PWM_POLARITY_NORMAL>;
}; 

red_pwm_led: red_pwm_led {
    /* use N output */
    pwms = <&pwm12 STM32_PWM_CHN(1) 4 PWM_POLARITY_NORMAL>;
}; 
```

`STM32_PWM_CHN` would add an _offset_ that indicates complementary output is used for channel 1 (up to 4 or 6 channels are available).

FRASTM commented:
This option is clear and readable to me because it direcly goes to the pwm driver
```
red_pwm_led: red_pwm_led {
    pwms = <&pwm12 1 4 (PWM_POLARITY_NORMAL | PWM_COMPLEMENTARY)>;
}; 
```

FRASTM committed:
dts: pwm : flag to enable a complementary output of a stm32 pwm channel

Depending on the stm32 soc and the timer instance, several channels
can enable the complementary output for the PWM signal
This flag completes the PWM_POLARITY for a PWM channel in the
upper byte of the pwm_flags_t.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
new version without impact on the pinctrl but only the PWM driver.

FRASTM commented:
splitting in 2 PR: one for the PWM API #42174
this one for using the #42174  with stm32 mcus

FRASTM commented:
Note that testing tests/drivers/pwm/pwm_loopback/ requires 32-bit timers that do not have complementary outputs.

FRASTM commented:
In case the PWM channel is already init'd, setting the polarity with LL_TIM_OC_SetPolarity depends on the complementary or not- complementary channel.

ABOSTM commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -103,6 +103,14 @@ static const uint32_t ch2ll[TIMER_MAX_CH] = {
 
+static const uint32_t ch2ll_n[TIMER_MAX_CH] = {
+	LL_TIM_CHANNEL_CH1N, LL_TIM_CHANNEL_CH2N,
+	LL_TIM_CHANNEL_CH3N, LL_TIM_CHANNEL_CH4,
```
Depending on STM32 series, CH4 can have a complementary channel,like on STM32G4, U5, ...
or not, like STM32L4, ...
Maybe macro IS_TIM_COMPLEMENTARY_CHANNELS  could be used.

ABOSTM commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -270,9 +278,17 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
 
 		oc_init.OCMode = LL_TIM_OCMODE_PWM1;
-		oc_init.OCState = LL_TIM_OCSTATE_ENABLE;
+
```
Few line above channel is defined like that 
`channel = ch2ll[pwm - 1u];`
it is then used in LL API like` LL_TIM_CC_DisableChannel()`, `LL_TIM_CC_IsEnabledChannel().`
But those API accept Complementary channel. So, is case of complementary, `ch2ll_n[]` should be used

ABOSTM commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -270,9 +278,17 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
-		oc_init.OCState = LL_TIM_OCSTATE_ENABLE;
+
+		/* the flags holds the PWM_COMPLEMENTARY information */
+		if ((flags & PWM_COMPLEMENTARY_MASK) == PWM_COMPLEMENTARY) {
```
An error could be raised in case channel doesn't have complementary capability, but it is set in DTS.

erwango replied:
@ABOSTM This is not possible as is today and proposed pinctrl bondings evolutions to enable it were rejected by mainatainer. This is specified as a "user's responsibility" to provide correct pinctrl configuration.

gmarull replied:
I guess capability is known, however, whether pin is correctly configured is not.

ABOSTM replied:
I mean runtime error thanks to macro IS_TIM_COMPLEMENTARY_CHANNELS()

gmarull reviewed (commented):
I think STM32 PWM bindings docs should be updated to explain how PWM nodes are expected to be configured and used.

henrikbrixandersen reviewed (changes requested):
Blocking for now as it adds the same changes as is being discussed in #42174.

FRASTM commented:
rebase
+ includes the LL_TIM_CHANNEL_CH4N  in the _ch2ll_n_ table when it exists ( stm32G474 for example)
+ warning at runtime when the channel has no complementary output but set by the DTS
+ use the complementary channel if needed in the LL_TIM_CC_DisableChannel and  LL_TIM_CC_IsEnabledChannel functions of the `pwm_stm32_pin_set`.

FRASTM committed:
dts: bindings : explains the flag for stm32 complementary channel

Depending on the stm32 mcus and the timer instance, several channels
can enable the complementary output for the PWM signal.
Example of a DTS <&pwm1 2 4 (PWM_POLARITY_NORMAL | PWM_COMPLEMENTARY)>;
Note that the timer channel must support the complementary output
on that channel : usually channels 1-3 + channel 4 on stm32g4x.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
changing to a similar approach  as to that for GPIO in https://github.com/zephyrproject-rtos/zephyr/pull/39767
That is a specific  PWM_STM32_COMPLEMENTARY flag for STM32 devices instead of a generic one.

FRASTM committed:
include: drivers: pwm: extend the pwm flags to be soc specific

Add 8 bits to the pwm_flags_t so that upper byte is reserved
for non-standard but soc specific dts flags and keeping the low
Byte for standard flags.
Some of SoC like STM32 mcus can then define their own flags
in their dt-bindings/pwm.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on include/dt-bindings/pwm/stm32_pwm.h:
```diff
@@ -0,0 +1,25 @@
+ * @{
+ */
+/** PWM complementary output pin is enabled */
+#define PWM_STM32_COMPLEMENTARY	(1 << 1)
```
OK to change pwm_flags_t in commit 59d5ba3721148079e6aed1e8a32a10c83ab090ec

FRASTM commented:
rebase
+ add commit 59d5ba3721148079e6aed1e8a32a10c83ab090ec to add 8bits to the pwm_flags_t  for soc specific flags

gmarull commented on dts/bindings/pwm/st,stm32-pwm.yaml:
```diff
@@ -31,4 +31,5 @@ pwm-cells:
   - channel
 # period in terms of nanoseconds
   - period
+# set 'PWM_STM32_COMPLEMENTARY' flag if the complementary output is selected
```
note: this comment is not part of the docs, so maybe update the binding description with some examples (it's just `STM32 PWM` now). This way we'll have stuff rendered in docs automatically.

gmarull reviewed (dismissed):
LGTM, one documentation suggestion.

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -259,20 +287,49 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
+	 */
+	if ((flags & PWM_STM32_COMPLEMENTARY_MASK) == PWM_STM32_COMPLEMENTARY) {
+		is_channel = ch2ll_n[pwm - 1u];
+#if !defined(LL_TIM_CHANNEL_CH4N)
+		if (pwm >= 4) {
+#else
+		if (pwm > 4) {
+#endif /* LL_TIM_CHANNEL_CH4N */
```
Let's define a STM32_MAX_N_CHANNEL constant and use it here.

FRASTM replied:
```
/** Maximum number of complemented timer channels. */
#define STM32_MAX_N_CHANNEL 4
```

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -259,20 +287,49 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
+#else
+		if (pwm > 4) {
+#endif /* LL_TIM_CHANNEL_CH4N */
+			LOG_WRN("Channel %d has NO complementary output", pwm);
```
Should it be LOG_ERR ?

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -259,20 +287,49 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
+	if ((flags & PWM_STM32_COMPLEMENTARY_MASK) == PWM_STM32_COMPLEMENTARY) {
+		is_channel = ch2ll_n[pwm - 1u];
+#if !defined(LL_TIM_CHANNEL_CH4N)
+		if (pwm >= 4) {
```
Shouldn't the check be done before `is_channel = ch2ll_n[pwm - 1u];` ?

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -227,6 +254,7 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
 	const struct pwm_stm32_config *cfg = dev->config;
 
 	uint32_t channel;
+	uint32_t is_channel; /* complementary output if used */
```
`is_channel` is not quite meaningful.
What about `channel_n_channel` ? (or anything better)

FRASTM replied:
proposal for `current_channel`

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -259,20 +287,49 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
 		LL_TIM_OC_StructInit(&oc_init);
 
 		oc_init.OCMode = LL_TIM_OCMODE_PWM1;
+#if defined(LL_TIM_CHANNEL_CH1N)
```
Break a line before #if defined and after #endif (not the other way round)

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -283,17 +340,20 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
 		}
 
 		LL_TIM_EnableARRPreload(cfg->timer);
+		/* in LL_TIM_OC_EnablePreload, the channel.is always the non-complementary */
```
`channel is`

FRASTM commented:
rebase on e8f93a915def87b9c7e1bce3323bec996d7863da
include the https://github.com/zephyrproject-rtos/zephyr/pull/42174
define STM32_MAX_N_CHANNEL
define a current_channel value

FRASTM commented:
rebase on 095001cdd12b100c1b89f91caf45fc17b357d2d2

FRASTM committed:
samples: blinky pwm running on the nucleo_l4r5zi

Adding an overlay to configure the nucleo_l4r5zi for testing the pwm
blinky application on red led (PB14 on nucleo board)
Each has a specific pwm output from different timers/channels

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
using stm32Cube LL definition to detect the Max nb of timer channels, to define the table of complementary channels
Raise an error if PWM_STM32_COMPLEMENTARY flag is set for a non-existing channel.

ABOSTM reviewed (dismissed):
LGTM

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -103,6 +89,20 @@ static const uint32_t ch2ll[TIMER_MAX_CH] = {
 #endif
 };
 
+#if defined(LL_TIM_CHANNEL_CH1N)
```
What about moving this #ifdef inside the array, so that array exist but has empty length if LL_TIM_CHANNEL_CH1N is not defined?
This remove the need to use `#ifdef LL_TIM_CHANNEL_CH1N` later in `pwm_stm32_pin_set()'

FRASTM replied:
done

erwango commented on dts/bindings/pwm/st,stm32-pwm.yaml:
```diff
@@ -26,9 +26,14 @@ properties:
+        - channel of the timer used for PWM
+        - period to set in ns
+        - flags : combination of standard flags like PWM_POLARITY_NORMAL
+        or specific flags like PWM_STM32_COMPLEMENTARY
```
A code snippet would be a plus:
```
		red_pwm_led: red_pwm_led {
			pwms = <&pwm1 2 4 (PWM_POLARITY_NORMAL | PWM_STM32_COMPLEMENTARY)>;
		};
```

FRASTM replied:
adding the example in the description

henrikbrixandersen commented:
Can #42174 be closed in favour of this PR?
